// tslint-config-security/dist/rules/tsrDetectPseudoRandomBytesRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var Lint = require("tslint");
var ts = require("typescript");
var Rule = (function (_super) {
    tslib_1.__extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        return this.applyWithFunction(sourceFile, walk);
    };
    Rule.metadata = {
        ruleName: 'tsr-detect-pseudo-random-bytes',
        description: 'Warns when crypto.pseudoRandomBytes() function is used',
        descriptionDetails: Lint.Utils.dedent(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["Any usage of crypto.pseudoRandomBytes() will trigger a warning.\n            See https://github.com/webschik/tslint-config-security#tsr-detect-pseudo-random-bytes"], ["Any usage of crypto.pseudoRandomBytes() will trigger a warning.\n            See https://github.com/webschik/tslint-config-security#tsr-detect-pseudo-random-bytes"]))),
        optionsDescription: '',
        options: null,
        type: 'functionality',
        requiresTypeInfo: false,
        typescriptOnly: false
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
function walk(ctx) {
    function visitNode(node) {
        if (node.kind === ts.SyntaxKind.PropertyAccessExpression) {
            var name = node.name;
            if (name && name.text === 'pseudoRandomBytes') {
                ctx.addFailureAtNode(node, 'Found crypto.pseudoRandomBytes which does not produce cryptographically strong numbers');
            }
        }
        return ts.forEachChild(node, visitNode);
    }
    return ts.forEachChild(ctx.sourceFile, visitNode);
}
var templateObject_1;
