// rollup-plugin-livereload/dist/index.cjs.js
'use strict';

var livereload$1 = require('livereload');
var path = require('path');

function livereload(options) {
  if ( options === void 0 ) options = { watch: '' };

  if (typeof options === 'string') {
    options = {
      watch: options
    };
  } else {
    options.watch = options.watch || '';
  }

  var enabled = options.verbose === false;
  var port = options.port || 35729;
  var server = livereload$1.createServer(options);

  // Start watching
  if (Array.isArray(options.watch)) {
    server.watch(options.watch.map(function (w) { return path.resolve(process.cwd(), w); }));
  } else {
    server.watch(path.resolve(process.cwd(), options.watch));
  }

  closeServerOnTermination(server);

  return {
    name: 'livereload',
    banner: function banner() {
      return ("(function(l, i, v, e) { v = l.createElement(i); v.async = 1; v.src = '//' + (location.host || 'localhost').split(':')[0] + ':" + port + "/livereload.js?snipver=1'; e = l.getElementsByTagName(i)[0]; e.parentNode.insertBefore(v, e)})(document, 'script');")
    },
    generateBundle: function generateBundle() {
      if (!enabled) {
        enabled = true;
        console.log(green('LiveReload enabled'));
      }
    }
  }
}

function green(text) {
  return '\u001b[1m\u001b[32m' + text + '\u001b[39m\u001b[22m'
}

function closeServerOnTermination(server) {
  var terminationSignals = ['SIGINT', 'SIGTERM'];
  terminationSignals.forEach(function (signal) {
    process.on(signal, function () {
      server.close();
      process.exit();
    });
  });
}

module.exports = livereload;
